torch/csrc/cuda/CUDAPluggableAllocator.cpp

```diff
@@ -0,0 +1,323 @@
+}
+
+// CUDAGraph interactions
+void CUDAPluggableAllocator::notifyCaptureBegin(
+    int device,
+    c10::cuda::CaptureId_t graph_id,
+    c10::cuda::MempoolId_t mempool_id) {
+  if (capture_begin_fn_.has_value()) {
+    capture_begin_fn_.value()(device, graph_id, mempool_id);
+  }
+}
+
+void CUDAPluggableAllocator::notifyCaptureAboutToEnd(
+    int device,
+    c10::cuda::CaptureId_t graph_id) {
+  if (capture_about_to_end_fn_.has_value()) {
+    capture_about_to_end_fn_.value()(device, graph_id);
+  }
+}
+
+void CUDAPluggableAllocator::notifyCaptureEnded(
+    int device,
+    c10::cuda::CaptureId_t graph_id) {
+  if (capture_ended_fn_.has_value()) {
+    capture_ended_fn_.value()(device, graph_id);
+  }
+}
+
+void CUDAPluggableAllocator::notifyCaptureDestroy(
+    int device,
+    c10::cuda::MempoolId_t mempool_id) {
+  if (capture_destroy_fn_.has_value()) {
+    capture_destroy_fn_.value()(device, mempool_id);
+  }
+}
+
+void CUDAPluggableAllocator::recordHistory(
+    bool enabled,
+    c10::cuda::CUDACachingAllocator::CreateContextFn context_recorder,
+    size_t alloc_trace_max_entries,
+    bool alloc_trace_record_context) {
+  TORCH_CHECK(
+      false,
+      "CUDAPluggableAllocator does not yet support recordHistory. "
+      "If you need it, please file an issue describing your use case.");
+}
+
+void CUDAPluggableAllocator::attachOutOfMemoryObserver(
+    c10::cuda::CUDACachingAllocator::OutOfMemoryObserver observer) {
+  TORCH_CHECK(
+      false,
+      "CUDAPluggableAllocator does not yet support attachOutOfMemoryObserver. "
+      "If you need it, please file an issue describing your use case.");
+}
+
+bool CUDAPluggableAllocator::needsPoolSpecificPeerAccess() {
+  return false;
+}
+
+std::string CUDAPluggableAllocator::name() {
+  return "pluggable";
+}
+
+std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator>
+    current_custom_allocator;
+
+std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator>
+getCurrentAllocator() {
+  if (current_custom_allocator)
+    return current_custom_allocator;
+  // Avoid freeing the main `CUDACachingAllocator` if it goes out of scope from
+  // python
+  return std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator>(
+      c10::cuda::CUDACachingAllocator::get(),
+      [](c10::cuda::CUDACachingAllocator::CUDAAllocator* allocator) {});
+}
+
+// TODO: add more functions in the argument
+std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator>
+createCustomAllocator(
+    std::function<void*(size_t, int, cudaStream_t)> alloc_fn,
+    std::function<void(void*, size_t, cudaStream_t)> free_fn) {
+  std::shared_ptr<CUDAPluggableAllocator> allocator(
+      new CUDAPluggableAllocator(alloc_fn, free_fn));
+  allocator->init(device_count);
+  return allocator;
+}
+
+void changeCurrentAllocator(
+    std::shared_ptr<c10::cuda::CUDACachingAllocator::CUDAAllocator> allocator) {
+  TORCH_CHECK(
+      !getCurrentAllocator()->initialized(),
```

Comment:
There is a race here where we can still end up with 

* Thread 1: getCurrentAllocator()
* Thread 2: allocate some memory with raw_alloc
* Thread 1: store new allocator
* Thread 2: raw_delete (calls wrong one and crashes)

However, I don't think this scenario is common. We should document that changing the allocator after allocations have occurred may result in undefined behavior. The best resolution to this longer term is to remove the raw_alloc/raw_delete interfaces.

Reply:
The pluggable allocator holds a history of active memory allocations, so it will hard error when it receives a pointer that it hasn't allocated. This list is needed to pass the `free` function the size and stream of the allocation as requested by the nvidia guys.